### ISSUE TITLE
Gradient Checkpointing

### ISSUE THREAD

**Issue by Dwrety** [kongds/E5-V#19]
Hi, in this line of your code https://github.com/kongds/E5-V/blob/main/ft_llm.py#L551

```
    if grad_checkpoint:
        model.enable_input_require_grads()
```

I have noticed you have enable the gradient on the embedding matrix. However, while saving the checkpoints, since you have wrapped the model with PEFTModel() class, the embedding matrix was not saved. I am wondering if this was designed on purpose? what are the benefits to enable the training on the embedding matrix?

**Comment by kongds** [comment]
Hi, this line does not indicate an unfrozen embedding; instead, it ensures that the input has requires_grad=True. You can see an explanation here.
https://github.com/huggingface/transformers/blob/0f67ba1d741d65b07d549daf4ee157609ce4f9c1/src/transformers/modeling_utils.py#L2273-L2278

**Comment by Dwrety** [comment]
> Hi, this line does not indicate an unfrozen embedding; instead, it ensures that the input has requires_grad=True. You can see an explanation here. https://github.com/huggingface/transformers/blob/0f67ba1d741d65b07d549daf4ee157609ce4f9c1/src/transformers/modeling_utils.py#L2273-L2278

Yes, and if you keep digging, you will find they actually unfreeze the embedding layer.
https://github.com/huggingface/transformers/blob/0f67ba1d741d65b07d549daf4ee157609ce4f9c1/src/transformers/modeling_utils.py#L1677
However, I have also checked for 1 iteration, that the embedding layer isn't altered after 1 gradient update. So this is just a very confusing implementation by HuggingFace.

**Comment by kongds** [comment]
This line enables gradients on the embedding output (e.g., `output = embedding(input_ids)`), but not on the embedding layer itself (the embedding weights will not be updated).

The `enable_input_require_grads` is useful with gradient checkpointing on PEFT. Since gradient checkpointing requires that the model's input has gradients enabled, this enables the input grads and prevents gradient computations from being None.

**Comment by Dwrety** [comment]
> This line enables gradients on the embedding output (e.g., `output = embedding(input_ids)`), but not on the embedding layer itself (the embedding weights will not be updated).
> 
> The `enable_input_require_grads` is useful with gradient checkpointing on PEFT. Since gradient checkpointing requires that the model's input has grads, this prevents gradient computations from being None.

I see, thank you.
